File: go/0388_longest_absolute_file_path.go
import "strings"

func lengthLongestPath(input string) int {
	res, cache, fileDirs := 0, make(map[int]int), strings.Split(input, "\n")

	for _, fileDir := range fileDirs {
		level := 0
		for fileDir[0] == '\t' {
			level, fileDir = level+1, fileDir[1:]
		}
		if level == 0 {
			if strings.Index(fileDir, ".") != -1 {
				res = len(fileDir)
			} else {
				cache[0] = len(fileDir)
			}
			continue
		}

		if strings.Index(fileDir, ".") != -1 { // file
			if curLen := cache[level-1] + 1 + len(fileDir); curLen > res {
				res = curLen
			}
		} else { // dir
			cache[level] = cache[level-1] + 1 + len(fileDir)
		}
	}
	return res
}

File: python/0388_longest_absolute_file_path.py
class Solution:
    def lengthLongestPath(self, input: str) -> int:
        stack = []
        curLength = 0
        maxLength = 0

        for path in input.split('\n'):
            level = 0
            i = 0
            while path[i:i + 1] == '\t':
                level += 1
                i += 1

            path = path[i:]  # remove prefix of '\t'
            while stack and len(stack) - 1 >= level:
                curLength -= len(stack[-1])
                stack.pop(-1)

            stack.append(path)
            curLength += len(path)

            if path.find('.') != -1 and curLength + len(stack) - 1 > maxLength:
                maxLength = curLength + len(stack) - 1
        return maxLength
